src/labels_transfer/pynndescent_knn/script.py

```diff
@@ -0,0 +1,110 @@
+import mudata as mu
+import numba
+import numpy as np
+from scipy.sparse import issparse
+import sys
+from pynndescent import PyNNDescentTransformer
+from sklearn.neighbors import KNeighborsClassifier
+from sklearn.pipeline import make_pipeline
+
+## VIASH START
+par = {
+    "input": "resources_test/pbmc_1k_protein_v3/pbmc_1k_protein_v3_mms.h5mu",
+    "modality": "rna",
+    "input_obsm_features": None,
+    "reference": "resources_test/annotation_test_data/TS_Blood_filtered.h5mu",
+    "reference_obsm_features": None,
+    "reference_obs_targets": ["cell_type"],
+    "output": "foo.h5mu",
+    "output_obs_predictions": None,
+    "output_obs_probability": None,
+    "output_uns_parameters": "labels_transfer",
+    "output_compression": None,
+    "weights": "uniform",
+    "n_neighbors": 15
+}
+meta = {
+    "resources_dir": "src/labels_transfer/utils"
+}
+## VIASH END
+
+sys.path.append(meta["resources_dir"])
+from helper import check_arguments, get_reference_features, get_query_features
+
+def setup_logger():
+    import logging
+    from sys import stdout
+
+    logger = logging.getLogger()
+    logger.setLevel(logging.INFO)
+    console_handler = logging.StreamHandler(stdout)
+    logFormatter = logging.Formatter("%(asctime)s %(levelname)-8s %(message)s")
+    console_handler.setFormatter(logFormatter)
+    logger.addHandler(console_handler)
+
+    return logger
+# END TEMPORARY WORKAROUND setup_logger
+
+def distances_to_affinities(distances):
+    stds = np.std(distances, axis=1)
+    stds = (2.0 / stds) ** 2
+    stds = stds.reshape(-1, 1)
+    distances_tilda = np.exp(-np.true_divide(distances, stds))
+
+    return distances_tilda / np.sum(distances_tilda, axis=1, keepdims=True)
```

Comment:
@VladimirShitov I'm noticing that with the current test data, this is returning all NaN for the probabilities (also the case in the previous component). It might be due to rounding/floating point precision. Do you have time to look into this? If not, can you provide me with a reference on which this functionality is based?

Reply:
Interesting, I didn't notice that! The function should work, I successfully used it in several projects as well as other people. It was suggested [in the scArches paper](https://www.nature.com/articles/s41587-021-01001-7) (see the "Cell type annotation" chapter in the methods), and is common approach in atlassing. Human Lung Cell Atlas, for example, leverages the same approach. Maybe something is wrong with the test data?

Reply:
So it looks like it's happening in the return statement `distances_tilda / np.sum(distances_tilda, axis=1, keepdims=True)`: In this case, there are many rows where the sum of the distances_tilda is zero (the standard deviations are very low, resulting causing `np.exp(-np.true_divide(distances, stds))` to approach zero. 

I've made an update to the normalization: if the sum of a row of the distance tilda equals 0, set the normalized values of that row all to one.

LMKWYT